Draft - Rh blockchain basics

rob-Hitchens opened the pull request:
Documentation content update for Module 0, all sections

This PR contains the content of "Module 0", which is meant as an introduction to blockchain technology for the participants of supported training programs.

### R1: Internal review (B9lab)

- Technical content review
  - [x] requested
  - [ ] completed
- Language review
  - [x] requested
  - [x] completed
- Technical platform review (remove if not required)
  - [x] requested
  - [ ] completed

### R2: External review

- Technical content review
  - [x] requested
  - [x] completed
- Language review
  - [ ] requested
  - [ ] completed

### R3: Internal QA review (B9lab)

- Technical content review
  - [ ] requested
  - [ ] completed
- Language review
  - [ ] requested
  - [ ] completed

### R4: Final external QA review

- Technical content review
  - [ ] requested
  - [ ] completed
- Language review
  - [ ] requested
  - [ ] completed

### RC: Release Candidate

- [x] Ready to be merged

CitMC commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
```diff
@@ -0,0 +1,218 @@
+
+This first module demystifies the foundations of blockchain technology. This background information will be important as you progress through the course. 
+
+The modules include learning materials for your studies, as well as examples and exercises for you to practice. You will also find freely accessible resources from other sites to help you dive deeper into the subject matter at hand.
```
The modules or the module? If you are referring to the modules in the whole Academy training, I would recommend: "The chapters in the Cosmos Developer Academy..." or "The Cosmos Developer Academy". If only the Introduction is meant, then I would recommend changing it to: "The chapter".

rob-Hitchens replied:
I'm talking about the _terminology_ (chapter? section?) that refers to everything in `0.0-B9lab-Blockchains` which is 5 pages of ambient knowledge that everyone should know before they attempt to unpack Cosmos.

CitMC replied:
Then I would go with chapter. ;-)

CitMC commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
I would recommend including a sentence explaining how Bitcoin is a decentralized and distributed ledger, as we only talked about distributed ledgers before and this might raise an eye brow.

rob-Hitchens replied:
At a high level, Bitcoin solves the double-spending problem by replacing the trusted, central ledger-keepers with a decentralized and distributed ledger that is maintained by a large network of ledger-keepers. Each member of the network has an exact replica of the ledger. In addition, no one can update the ledger without establishing consensus with the other ledger-keepers.

CitMC commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
Should COMMIT and ROLLBACK be in code tags (``)?

rob-Hitchens replied:
Not exactly. I think I'll italicize. 

Syntax depends on your database implementation, e.g. MS-SQL, MySQL, Oracle. The _concept_ is to enforce atomicity at the database level.

CitMC commented on academy/0.0-B9lab-Blockchains/2_public.md:
```diff
@@ -0,0 +1,178 @@
+
+![](images/timeline.png)
+
+The Tendermint Consensus solution was announced in 2014.
```
I would include more information here if you already want to talk about Tendermint. But in my review, I moved the image around a bit to a later point because I believe it fits more seamlessly.

CitMC commented on academy/0.0-B9lab-Blockchains/2_public.md:
```diff
@@ -0,0 +1,178 @@
+
+Let us start by diving into public blockchains and their most known examples, Bitcoin and Ethereum.
+
+## Public Blockchains
```
I would strongly recommend moving the content on public blockchains into the next section to focus this section on one aspect. Recommendation is implemented in review changes.

CitMC commented on academy/0.0-B9lab-Blockchains/2_public.md:
```diff
@@ -0,0 +1,178 @@
+
+The most obvious way of operating blockchain protocols comes in the form of a public network. This is what blockchain technology was originally invented for and remains arguably its most powerful use.
+
+![](images/00_02_p2p_network_dark.png)
```
Why not rather use a tailored version of 00_08_public_vs_private_comparison-01.png that doesn't display managed networks and include the current version of 00_08_public_vs_private_comparison-01.png later when you introduce managed networks?

rob-Hitchens replied:
I'm not sure I "get it". If you mean cropping this section from the diagram that appears later on, I'm good with that. 

![image](https://user-images.githubusercontent.com/26111999/158899671-58ba5403-b754-4813-a178-c103f2867f42.png)

rob-Hitchens replied:
@CitMC Do I understand your suggestion?

CitMC commented on academy/0.0-B9lab-Blockchains/2_public.md:
```diff
@@ -0,0 +1,178 @@
+
+<ExpansionPanel title="Faster Block Time">
+
+Ethereum has a much faster block time, (currently around 15s), made practical thanks to its implementation of the GHOST protocol [[1]](http://www.cs.huji.ac.il/~yoni_sompo/pubs/15/inclusive_full.pdf. Ethereum implements a variation of Bitcoin's proof-of-work consensus algorithm called Ethash, which is intended to be ASIC-resistant and GPU friendly[[2]](https://github.com/ethereum/wiki/wiki/Ethash ...
```
The link is not working, please recheck.

CitMC commented on academy/0.0-B9lab-Blockchains/3_managed.md:
```diff
@@ -0,0 +1,41 @@
+# Introduction to Managed Networks
+
+Managed networks, just like public networks, rely on blockchain data structures. But unlike public blockchain networks they do not *necessarily* need to mitigate the Byzantine General's Problem because they operate in a predictable environment with elements of authority, hierarchy and accountability.
+
```
Completely missing here in my opinion is an introduction of what managed networks are, as you dive right into the comparison. What do we mean by private/managed?

rob-Hitchens replied:
## Managed Blockchain Networks

Managed networks, just like public networks, rely on blockchain data structures. Unlike public blockchain networks, they do not necessarily need to mitigate the Byzantine Generals' Problem because they operate in a predictable environment with elements of authority, hierarchy, and accountability.

Managed networks can be used in cases where elements of trust already exist between the participants. For example, consider a network of financial institutions. They could use blockchain technology to settle inter-bank transactions. In that case, there would be no need for public access and, indeed, public access would be undesirable. Instead, they might consider a private network with each institution operating exactly one validator. All participants would understand that no operator in a single institution would be able to corrupt the network. In fact, it would be much easier for them, as a group, to alter a history if they found it detrimental since it would only require the coordinated action of a small number of known validators. 

Managed networks are typically governed through traditional governance processes that are appropriate for the shared goals of the participants.

CitMC commented on academy/0.0-B9lab-Blockchains/3_managed.md:
```diff
@@ -0,0 +1,41 @@
+
+- **can be designed for a limited number of vetted and approved participants.** Accommodation of performance-challenged and poorly connected nodes is of lesser importance. 
+
+- **can be designed for optimised performance.** Most participants in an Enterprise network will be well-capable of standing up well-connected high-performance and high-availability nodes. A group of participants can agree to raise the bar defining minimum system requirements significantly. For example, participation might be limited to sizable Enterprise-class servers with redundant low-latency, ...
```
I'm not sure we need to include a rather long elaboration -starting with "For example, participation...". This makes the bullet point rather long. Maybe you could dive into it in a paragraph following the bullet points.

CitMC commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,112 @@
+
+An extension of proof-of-stake algorithms is called delegated-proof-of-stake (DPoS). It is used for example in BitShares, EOS, Steem, Lisk and Tron.
+
+<HighlightBox type="info">
```
Deleted this highlight box because I would either include a short introduction to all projects mentioned in the previous sentence or none.

CitMC commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,112 @@
+
+PBFT was published in 1999 and arose from academia. It is a three-phase protocol, in which the client sends a request to a so-called primary. 
+
+In the first phase, the primary broadcasts the request with a sequence number to the replicas. Then the replicas agree on the sequence number and create a message. 
```
What is a replica? I would introduce a sentence or expand the sentence to explain very briefly the concept.

CitMC commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,112 @@
+
+[Tendermint](./1-what-is-cosmos/blockchain-and-cosmos.md) will be explored in detail later on. 
+
+## Many others ##
```
I moved this part in my commits to the beginning of the section to give it a seamless flow overall.

CitMC commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,112 @@
+
+## Many others ##
+
+Many teams have experimented with distributed consensus using many algorithms deployed to various chains. In the context of blockchain technology, "consensus" refers to a method of ordering transactions. The many algorithms that exist can be compared in terms of the distribution of authority, expected performance and transaction finality. Transaction finality can be categorized as "probablistic"  ...
```
I shortened this paragraph to avoid repetitions.

CitMC commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,112 @@
+
+The above are the most popular by usage as well as the most important to keep in mind when diving into consensus in Cosmos. 
+
+<HighlightBox type="reading">
```
I tailored the reading list and deleted all readings that seem to not be completely helpful for those diving into Cosmos and added some readings that would fit in my opinion. Please recheck.

rob-Hitchens replied:
fantastic. thank you.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+tag: fast-track
+---
+
+# Fundamentals
```
Only include a headline level if there is at least one sentence between headlines. Please include an introductory sentence after this main headline.

CitMC replied:
I believe I found a solution. Please recheck the headlines in my commit.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+
+We can proceed with the understanding that signed messages from Alice could only come from someone with knowledge of Alice's private key and messages that are encrypted for Bob can only be deciphered by Bob. 
+
+![](images/00_13_mix_n_match_keys_v3.png)
```
Moved the image further up because I feel it would better fit into the content before diving into the recap "What is this sorcery?"

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+
+#### Mix and match
+
+It is possible to mix both ideas, whereby Alice encrypts her message with Bob's public key, then signs the encrypted file with her private key. Upon reception, Bob verifies the signature with Alice's public key, to be sure it came from Alice, then decrypts the file with his private key.
```
I made a list out of it to be more consistent in regard to the two previous scenario examples.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+
+![](images/00_13_mix_n_match_keys_v3.png)
+
+### Key management, PKI
```
I changed the heading level a bit. Please review.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+
+It is possible to index content by its hash, in essence creating a hashtable. If you have used IPFS or BitTorrent's magnet links, among others, then you have used a hashtable.
+
+### Digital Certificates
```
I would recommend deleting this whole sub-section.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+
+<HighlightBox type="reading">
+
+		* Zero-knowledge proof [https://en.wikipedia.org/wiki/Zero-knowledge_proof](https://en.wikipedia.org/wiki/Zero-knowledge_proof)
```
Deleted this further reading as 0-knowledge-proof is not introduced once in the content.

CitMC commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,180 @@
+<HighlightBox type="reading">
+
+		* Zero-knowledge proof [https://en.wikipedia.org/wiki/Zero-knowledge_proof](https://en.wikipedia.org/wiki/Zero-knowledge_proof)
+		* The Mathematics of the RSA Public-Key Cryptosystem [http://www.mathaware.org/mam/06/Kaliski.pdf](http://www.mathaware.org/mam/06/Kaliski.pdf)
```
The link provided leads to a 404.

AndrewB9lab commented on academy/0.0-B9lab-Blockchains/3_managed.md:
```diff
@@ -0,0 +1,134 @@
+
+The precise mechanism employed by Ethereum is called GHOST. GHOST includes so-called uncles that are propagated into the network too late to rise to the level of network consensus. This increases the total difficulty of the chain, makes smaller block times possible, and rewards miners of uncles for contributing to the overall strength of the network.
+
+Ethereum's solution to achieve faster block times draws attention to the universal challenge of blockchain technology, performance, and throughput. Cosmos solves for performance in two ways: with a faster consensus process, and by introducing parallelism.
```
@xavierlepretre , I query the sentence **Ethereum's solution to achieve faster block times draws attention to the universal challenge of blockchain technology, performance, and throughput.**

It is ambiguous whether there are:

1.  **three** challenges (blockchain technology, performance, and throughput); or 
2. **two** challenges _of blockchain_ (performance, and throughput).

xavierlepretre replied:
It is 2 challenges.

AndrewB9lab commented on academy/0.0-B9lab-Blockchains/3_managed.md:
```diff
@@ -0,0 +1,134 @@
+
+A well-understood challenge related to reduced block time stems from network latency. Since nodes do not learn about newly discovered blocks simultaneously, at any given time a portion of the miners will be working on already solved (i.e. old) blocks. If these miners find a solution, they might not be rewarded and a "fork" of the chain results, in which rival chains with a common true history spr ...
+
+![Uncles in Ethereum](images/00_18_block_uncles-01.png)
```
Not directly a "language" review factor, but: as a layperson, I find the depiction of the blockchain/uncles dynamic in the graphic _00_18_block_uncles-01.png_ a bit counter-intuitive. It seems to me that the direction of the arrows AND the atypical right-to-left orientation to represent old-to-new could provoke confusion.

I would clarify by expanding the labels "Block" and "Uncle" to indicate newness, and orient the "aging" of the system to show older at the left and newer at the right (see attached image).

![Blocks-and-uncles](https://user-images.githubusercontent.com/104000630/165515303-f1ac5725-4d8b-4c7e-b0ee-326ccdf13f5c.png)

I have maintained the use of arrows which point towards the source of a block's history. However, I would observe that this also seems counter-intuitive, with diagrams typically pointing _from_ one state _toward_ a subsequent state. Perhaps the usage here is a convention in blockchain specifically, but if not I would be inclined to reverse the arrow directions in all instances here.

xavierlepretre replied:
It is an issue with which we often grapple. Personally, I would add a _passage of time_ arrow below all going from left to right, the `x` axis. Perhaps with the legend of "time of appearance on network" or suchlike.
Then have each block put alongside the `x` axis as is visible in the 2 versions of the diagram.
The arrows would point leftward, but I would make them tenuous, or dotted, with a legend of: A <- B: A is parent block of B.
And name _Uncle B_ a block that appears later than _Block B_ but is then included in _Block C_, not _Block A_. The arrow would point from _Block C_ to _Uncle B_ with another tenuous arrow type and the legend A <- B: A is an uncle of B.
Keep in mind that this uncle concept is non-existent in PoS blockchains. So we could just mention it in passing with a link.
@rob-Hitchens

rob-Hitchens replied:
I think the passage of time arrow is a good idea to possibly add in the future but not crucial for this stage. We've used this image before and it's just in passing mention, in this case, to help position Cosmos against the competition. The complexity of this process is one of the elements Tendermint (arguably) does better with validator sets and true transaction finality.

julienrbrt commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,138 @@
+</HighlightBox>
+
+<!-- TODO: Check this markdown format -->
+<!-- TODO: THIS VIDEO IS CURRENTLY PRIVATE ON YOUTUBE -->
```
It is just unlisted now

julienrbrt commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,138 @@
+
+For any given block, a validator is selected in a pseudo-random fashion. The validator with a larger stake has a higher probability of being selected to generate a block. While PoS systems generally reward validators with new coins for honest behavior, so-called **block rewards**, validators also receive transaction fees in return for generating blocks that the rest of the network accepts.
+
+## Delegated-Proof-of-Stake (DPoS)
```
This can be nice that mention the little difference between DPoS and PoA

julienrbrt commented on academy/0.0-B9lab-Blockchains/5_crypto.md:
```diff
@@ -0,0 +1,186 @@
+
+### A closer look at a hash function
+
+MD5 is such a hash function:
```
It is better to not use a hash function that has been proven to not fulfill the description of above.

julienrbrt reviewed (approved):
Few comments, feel free to take consideration of them or not. It is mostly suggestion or addition.
So approving. :)

xavierlepretre commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
Don't we mention elsewhere in the content that MD5 has been broken? It should be close by if missing. I think there is value in telling that a hash can be broken. Not sure it would be better told if we started with one that is not broken.

rob-Hitchens replied:
I've change the example to https://www.browserling.com/tools/all-hashes

xavierlepretre commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
You mean Merkle tree?

xavierlepretre commented on academy/0.0-B9lab-Blockchains/1_blockchain.md:
It could be added IMO.

rob-Hitchens commented:
Resolved comments with some minor edits. Unfurl the resolved conversations for area of interest. 

The most noteworthy edits are:

1. Changed the hash toy example. Not (just) MD5 anymore. 
2. Reorganized the Ethereum fast-blocks description. I think it flows better now, and the redundant statement is gone. 
3. Added a little blurb (little!) about PoA. 
4. Removed the video link per the planning session discussion on this topic.

xavierlepretre commented on academy/0.0-B9lab-Blockchains/4_consensus.md:
```diff
@@ -0,0 +1,135 @@
+<HighlightBox type="info">
+
+The idea of substantiating a claim through an arbitrary amount of work was previously suggested as a way to combat denial-of-service (DoS) attacks and spam in other contexts. "Work" is understood as the amount of specific computational effort used.
+
```
```suggestion

</HighlightBox>

```

xavierlepretre commented on academy/0.0-B9lab-Blockchains/3_managed.md:
```diff
@@ -0,0 +1,142 @@
+
+Another way of understanding this is to consider that at any point in time, a predictable percentage of the total mining hardware in the network will be working on the wrong chain, because of latency, and this percentage increases with shorter block times. It is not beneficial to the network or the miners to waste hardware resources that are intended to secure the network. 
+
+While Bitcoin addresses this efficiency issue with a relatively long block time of 10 minutes, Ethereum addresses this concern with a partial reward strategy: valid blocks that aren't ultimately included in the canonical chain (because another chain becomes longer) are still included on the side and known as **uncles**. Miners of uncles receive a smaller reward than regular block miners. 
```
```suggestion
While Bitcoin addresses this efficiency issue with a relatively long block time of 10 minutes, Ethereum addresses this concern with a partial reward strategy: valid blocks that aren't ultimately included in the canonical chain (because another chain becomes longer) can still be included on the side and known as **uncles**. Miners of uncles receive a smaller reward than regular block miners. 
```
